### ISSUE TITLE
[Question] Dynamic linking doesn't emit link args to rustc from build.rs?

### ISSUE THREAD

**Issue by tiras-j** [fede1024/rust-rdkafka#109]
Heya Federico, hope things are going well :)

I was wondering for the dynamic linking option how the build script works without emitting `cargo:rustc-link-lib=<library>` (e.g. [cargo build scripts docs](https://doc.rust-lang.org/cargo/reference/build-scripts.html#outputs-of-the-build-script)). Is there something special happening via the pkg-config probe?

Just curious because I'm dealing with some dynamic system library linking myself and was wondering what blackmagic you knew and whether you'd be willing to teach it!

**Comment by flavray** [comment]
Hey Josh!

tl;dr: It looks like the pkg-config probe is indeed emitting it for us, thanks to the [cargo_metadata](https://docs.rs/pkg-config/0.3.14/pkg_config/struct.Config.html#method.cargo_metadata) flag

```
17:01:53 flavr@alpha rdkafka-sys master
>> CARGO_FEATURE_DYNAMIC_LINKING=1 cargo build
   Compiling libc v0.2.45
   Compiling pkg-config v0.3.14
   Compiling cc v1.0.26
   Compiling libz-sys v1.0.25
   Compiling num_cpus v0.2.13
   Compiling rdkafka-sys v0.11.6-1 (/Users/flavr/Documents/Personal/rust-rdkafka/rdkafka-sys)
    Finished dev [unoptimized + debuginfo] target(s) in 9.20s

17:02:03 flavr@alpha rdkafka-sys master
>> cat target/debug/build/rdkafka-sys-f6183bc0e654512a/output
cargo:rustc-link-search=native=/usr/local/lib
cargo:rustc-link-lib=rdkafka
```

-----

non-tl;dr: (I have close-to-zero knowledge about pkg-config, so I got curious 😄)
Looking a bit more into the pkg_config [.probe](https://github.com/rust-lang/pkg-config-rs/blob/2890031fc9e22c2dfd97113f53ed1f5abcb7b1ad/src/lib.rs#L309), it:

- Runs `pkg-config --libs --cflags rdkafka`, which gives

```
17:07:47 flavr@alpha ~
>> pkg-config --libs --cflags rdkafka
-I/usr/local/include -L/usr/local/lib -lrdkafka
```

- Parses the output and finds the [`-lrdkafka`](https://github.com/rust-lang/pkg-config-rs/blob/2890031fc9e22c2dfd97113f53ed1f5abcb7b1ad/src/lib.rs#L494)
- Outputs the `cargo:rustc-link-lib=rdkafka` [meta](https://github.com/rust-lang/pkg-config-rs/blob/2890031fc9e22c2dfd97113f53ed1f5abcb7b1ad/src/lib.rs#L505)

So the build script does output the metadata for cargo, only implicitly thanks to pkg-config's magic

(Federico does not maintain this repo anymore, I'm around to help maintain it. But he did help me find the answer for your question 🙂)

Hope this helps!

**Comment by tiras-j** [comment]
Hey Flavien!

That's awesome! I think I'm going to be using pkg-config myself going forward, since that should probably help mitigate some unusual linking situations anyways, with some extra magic to boot!

Thanks for the super detailed feedback. Hope everything is going well!
